src/ompl/geometric/planners/rrt/src/RRTstar.cpp

```diff
@@ -573,73 +652,322 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
     }
 }
 
-int ompl::geometric::RRTstar::pruneTree(const base::Cost pruneTreeCost)
+int ompl::geometric::RRTstar::pruneTree(const base::Cost& pruneTreeCost)
 {
-    const int tree_size = nn_->size();
-    pruneScratchSpace_.newTree.reserve(tree_size);
-    pruneScratchSpace_.newTree.clear();
-    pruneScratchSpace_.toBePruned.reserve(tree_size);
-    pruneScratchSpace_.toBePruned.clear();
-    pruneScratchSpace_.candidates.clear();
-    pruneScratchSpace_.candidates.push_back(startMotion_);
-    std::size_t j = 0;
-    while (j != pruneScratchSpace_.candidates.size())
+    // Variable
+    // The percent improvement (expressed as a [0,1] fraction) in cost
+    double fracBetter;
+    // The number pruned
+    int numPruned = 0;
+
+    if (opt_->isFinite(prunedCost_))
     {
-        Motion *candidate = pruneScratchSpace_.candidates[j++];
-        if (opt_->isCostBetterThan(pruneTreeCost, costToGo(candidate)))
-            pruneScratchSpace_.toBePruned.push_back(candidate);
-        else
+        fracBetter = std::abs((pruneTreeCost.value() - prunedCost_.value())/prunedCost_.value());
+    }
+    else
+    {
+        fracBetter = 1.0;
+    }
+
+    if (fracBetter > pruneThreshold_)
+    {
+        // We are only pruning motions if they, AND all descendents, have a estimated cost greater than pruneTreeCost
+        // The easiest way to do this is to find leaves that should be pruned and ascend up their ancestry until a motion is found that is kept.
+        // To avoid making an intermediate copy of the NN structure, we process the tree by descending down from the start(s).
+        // In the first pass, all Motions with a cost below pruneTreeCost, or Motion's with children with costs below pruneTreeCost are added to the replacement NN structure,
+        // while all other Motions are stored as either a 'leaf' or 'chain' Motion. After all the leaves are disconnected and deleted, we check
+        // if any of the the chain Motions are now leaves, and repeat that process until done.
+        // This avoids (1) copying the NN structure into an intermediate variable and (2) the use of the expensive NN::remove() method.
+
+        // Variable
+        // The queue of Motions to process:
+        std::list<Motion*> motionQueue;
+        // The list of leaves to prune
+        std::list<Motion*> leavesToPrune;
+        // The list of chain vertices to recheck after pruning
+        std::list<Motion*> chainsToRecheck;
```

Comment:
No, I didn't benchmark this against the previous scheme as the previous pruning method was inadmissible. By removing a vertex independent of the heuristic-value of its children, the pruning method does not maintain uniform density in the subproblem, and more importantly can remove vertices that could be part of a better solution (e.g., in the attached picture, the highlighted vertex and its descendents would be removed by the old method, even though some of the descendents will be part of a better solution once a passage is found through the gap).

![output](https://cloud.githubusercontent.com/assets/5250910/9202206/c009a366-401e-11e5-8fb5-e5c2aaff9301.png)


Regardless, you are correct that in the previous method that `std::vector` was allocated in a single call, but the difference is that it did so twice, both of the size of the NN structure (it made `std::vector`s for the vertices to keep and prune, respectively).

Alternatively, this method does not store the vertices to keep in a `std::vector`, but instead immediately adds them back to the NN structure. So no allocation for those vertices ever occurs. The other method didn't do this because it was actually counting the vertices to decide whether to go through with the prune or not (see other comment).

The only vertices it does store are ones that it cannot immediately make a decision on (`chainsToRecheck`) and those that are to be pruned (`leavesToPrune`). For these vertices, it's just using the `std::list` as a queue...

I guess the underlying assumption is that you will be pruning (significantly) less vertices than the size of the graph. Would preallocating to the entire size of the tree still be beneficial? I don't really know.